watchface

```diff
@@ -0,0 +1,406 @@
+    fi
+}
+
+function pullFiles {
+    local user="$1"
+    local sourcedir="$2"
+    local destdir="$3"
+    if [ "$ADB" = "true" ] ; then
+        adb pull ${sourcedir} "${destdir}"
+    else
+        scp -P"${WATCHPORT}" -r "${user}@${WATCHADDR}:${sourcedir}" "${destdir}"
+    fi
+}
+
+function pushWatchface {
+    pushFiles "root" "${1}"'/usr/share/*' "/usr/share/"
+}
+
+function pushWallpaper {
+    local source="$1"
+    local wallpaper
+    wallpaper="$(basename "$1")"
+    local destination="/usr/share/asteroid-launcher/wallpapers/full/${wallpaper}"
+    pushFiles "root" "${source}" "${destination}"
+    setDconf "/desktop/asteroid/background-filename" "${destination}"
+}
+
+function restartCeres {
+    doWatchCommand "root" "systemctl restart user@1000"
+}
+
+function rebootWatch {
+    doWatchCommand "ceres" "reboot"
+}
+
+function activateWatchface {
+    setDconf "/desktop/asteroid/watchface" "/usr/share/asteroid-launcher/watchfaces/${1}.qml"
+}
+
+function setTimeZone {
+    doWatchCommand "root" "timedatectl set-timezone $1"
+}
+
+function forgetKeys {
+    ssh-keygen -R "${WATCHADDR}"
+    ssh-keygen -R watch
+}
+
+function watchfaceClone {
+    local files
+    mapfile -t files <<<"$(find . -path './'"$1"'/*' -type f | awk '{ dst=$0 ; gsub(/'"$1"'/,"mmmm",dst);  print $0 " " dst }')"
+    for myfile in "${files[@]}" ; do
+        IFS=" " read -r -a onefile <<< "${myfile}"
+        #echo "Copying from ${onefile[0]} to ${onefile[1]}"
+        destdir="${onefile[1]%/*}"
+        extension="${onefile[1]##*.}"
+        mkdir -p "${destdir}"
+        cp "${onefile[0]}" "${onefile[1]}"
+        if [ "${extension}" = "qml" ] ; then
+            sed -i "s/$1/$2/g" "${onefile[1]}"
+        fi
+    done
+}
+
+function deployface {
+    # trim trailing / if present
+    local face="${1%/*}"
+    local activate="$2"
+    if [ -d "${face}/usr/share/asteroid-launcher/watchfaces/" ] ; then
+        echo "Deploying ${face}"
+        pushWatchface "${face}"
+        if [ "${activate}" = "true" ] ; then
+            activateWatchface "${face}"
+            if [ -n "${WALLPAPER}" ] ; then
+                pushWallpaper "${WALLPAPER}"
+            fi
+            restartCeres
+        fi
+    else
+        echo "Error: ${face} does not appear to be a watchface"
+    fi
+}
+
+function testface {
+    # trim trailing / if present
+    local face="${1%/*}"
+    if [ -d "${face}/usr/share/asteroid-launcher/watchfaces/" ] ; then
+        echo "Testing ${face}"
+        if [ -n "${WALLPAPER}" ] ; then
+            qmlscene "${face}" "${WALLPAPER}" loader.qml
+        else
+            qmlscene "${face}" "background.jpg" loader.qml
+        fi
+    else
+        echo "Error: ${face} does not appear to be a watchface"
+    fi
+}
+
+function cloneface {
+    # trim trailing / if present
+    local face="${1%/*}"
+    if [ -d "${face}/usr/share/asteroid-launcher/watchfaces/" ] ; then
```

Comment:
Let's extract this logic into a helper function that verifies a watchface is valid and returns the sanitized name on success and exit in case of error.

Then functiony mc functionfaces don't have to repeat the same code and can just consider the watchface valid after the call. That means you can also deindent the logic of the code.

Generally speaking, the following pattern is preferred :
```
if (error)
  return;

doStuff
```

over this pattern:
```
if (!error) {
  doStuff
} else {
  return
}
```

because when new error conditions are added the control flow is easier to follow.

```
if (error1)
   return
if (error2)
  return

doStuff
```

rather than

```
if (!error1) {
  if (!error2) {
     doStuff
  } else {
    return
  }
} else {
  return
}
```

Reply:
I've broken out the common code into a new function named `sanitizeSourceWatchface`.